### PR TITLE
Dockerfile Run COPY and CHOWN as root

### PR DESCRIPTION
**Pull request by adambkaplan**
Ensure all chown commands run as root.
Assemble user in general does not have permission to do this.

### CODE REVIEW

**Comment by adambkaplan** [comment]
/assign @bparees 

@openshift/sig-developer-experience fyi

File: pkg/build/strategies/dockerfile/dockerfile.go [hunk 1]
```diff
@@ -139,13 +139,12 @@ func (builder *Dockerfile) CreateDockerfile(config *api.Config) error {
...
 		buffer.WriteString(fmt.Sprintf("FROM %s as cached\n", imageTag))
-		if len(config.AssembleUser) > 0 {
-			buffer.WriteString(fmt.Sprintf("USER %s\n", imageUser))
-		}
 		var artifactsScript string
 		if _, provided := providedScripts[constants.SaveArtifacts]; provided {
+			// Switch to root to COPY and chown content
 			glog.V(2).Infof("Override save-artifacts script is included in directory %q", builder.uploadScriptsDir)
 			buffer.WriteString("# Copying in override save-artifacts script\n")
+			buffer.WriteString("USER root\n")
```
Comment (adambkaplan) [inline pkg/build/strategies/dockerfile/dockerfile.go]: use `USER 0` instead?

File: pkg/build/strategies/dockerfile/dockerfile.go [hunk 2]
```diff
@@ -154,41 +153,31 @@ func (builder *Dockerfile) CreateDockerfile(config *api.Config) error {
...
 			first = false
 		}
-		for k, v := range config.Labels {
-			if !first {
-				buffer.WriteString(fmt.Sprintf(" \\\n      "))
-			}
-			buffer.WriteString(fmt.Sprintf("%q=%q", k, v))
-			first = false
-		}
-
```
Comment (adambkaplan) [inline pkg/build/strategies/dockerfile/dockerfile.go]: Noted that if `config.Labels` had duplicate keys, we would have the same ENV variable set twice.

File: pkg/build/strategies/dockerfile/dockerfile.go (hunk 1 above)
Comment (bparees) [reply pkg/build/strategies/dockerfile/dockerfile.go]: this should be fine i think, especially given our intent to remove it.

File: pkg/build/strategies/dockerfile/dockerfile.go [hunk 3]
```diff
@@ -154,41 +153,31 @@ func (builder *Dockerfile) CreateDockerfile(config *api.Config) error {
...
+		// Switch to the image user if it is not root
+		if len(imageUser) > 0 && imageUser != "root" {
+			buffer.WriteString(fmt.Sprintf("USER %s\n", imageUser))
+		}
 		buffer.WriteString(fmt.Sprintf("RUN if [ -s %[1]s ]; then %[1]s > %[2]s; else touch %[2]s; fi\n", artifactsScript, artifactsTar))
 	}
 
 	buffer.WriteString(fmt.Sprintf("FROM %s\n", config.BuilderImage))
+	// Run as root to COPY and chown source content
+	buffer.WriteString("USER root\n")
```
Comment (bparees) [inline pkg/build/strategies/dockerfile/dockerfile.go]: can we move this closer to the copy?  right now it looks like it'll come out before the label+env commands.

File: pkg/build/strategies/dockerfile/dockerfile.go (hunk 2 above)
Comment (bparees) [reply pkg/build/strategies/dockerfile/dockerfile.go]: you mean the same LABEL twice?

that's ok, the latter takes precedence i believe, so as long as we set the user's value last so they can overwrite our generated label, it's fine.

Comment (bparees) [reply pkg/build/strategies/dockerfile/dockerfile.go]: debating whether it's better as it was so we can see the values we generated in the dockerfile, or better the way you have it..   i guess i can go either way, so let's go with this.  less layers anyway.

**Comment by adambkaplan** [comment]
@bparees updated per comments

**Comment by adambkaplan** [comment]
[test]

**Comment by bparees** [comment]
no offense but i can't wait to undo this :)

[merge]

**Pull request merged**